Many Pex tests fail under Python 2.7 in CI

I was finally able to reproduce this locally by re-creating my tox venv which, as a side effect, upgraded virtualenv. The latter is the problem.

The errors look like:
```
$ tox -epy27 -- -vvsk test_entry_points_dict
.package create: /home/jsirois/dev/pantsbuild/jsirois-pex/.tox/.package
.package installdeps: flit_core >=2,<3
py27 create: /home/jsirois/dev/pantsbuild/jsirois-pex/.tox/py27
py27 installdeps: pytest==4.6.6, mock==3.0.5
py27 inst: /home/jsirois/dev/pantsbuild/jsirois-pex/.tox/.tmp/package/1/pex-2.1.9.tar.gz
py27 installed: DEPRECATION: Python 2.7 reached the end of its life on January 1st, 2020. Please upgrade your Python as Python 2.7 is no longer maintained. A future version of pip will drop support for Python 2.7. More details about Python 2 support in pip, can be found at https://pip.pypa.io/en/latest/development/release-process/#python-2-support,atomicwrites==1.4.0,attrs==19.3.0,configparser==4.0.2,contextlib2==0.6.0.post1,funcsigs==1.0.2,importlib-metadata==1.6.0,mock==3.0.5,more-itertools==5.0.0,packaging==20.3,pathlib2==2.3.5,pex==2.1.9,pluggy==0.13.1,py==1.8.1,pyparsing==2.4.7,pytest==4.6.6,scandir==1.10.0,six==1.14.0,wcwidth==0.1.9,zipp==1.2.0
py27 run-test-pre: PYTHONHASHSEED='1467207441'
py27 run-test: commands[0] | pytest --ignore=tests/test_integration.py -vvsk test_entry_points_dict
==================================================================================================================================== test session starts =====================================================================================================================================
platform linux2 -- Python 2.7.18, pytest-4.6.6, py-1.8.1, pluggy-0.13.1 -- /home/jsirois/dev/pantsbuild/jsirois-pex/.tox/py27/bin/python
cachedir: .tox/py27/.pytest_cache
rootdir: /home/jsirois/dev/pantsbuild/jsirois-pex
collected 179 items / 178 deselected / 1 selected                                                                                                                                                                                                                                            

tests/test_bdist_pex.py::test_entry_points_dict running bdist_pex
Writing my_app to /tmp/tmpsgDVu1/my_app-0.0.0.pex
Failed to create pex via /home/jsirois/dev/pantsbuild/jsirois-pex/.tox/py27/bin/python -s -m pex /tmp/tmpxEPsaM --output-file /tmp/tmpsgDVu1/my_app-0.0.0.pex --script my_app:
ERROR: Exception:
Traceback (most recent call last):
  File "/home/jsirois/.pex/pip.pex/3616f444e6841b3bb5e495d304f6de6a76f743d8/.deps/pip/pip/_internal/cli/base_command.py", line 172, in _main
    status = self.run(options, args)
  File "/home/jsirois/.pex/pip.pex/3616f444e6841b3bb5e495d304f6de6a76f743d8/.deps/pip/pip/_internal/commands/install.py", line 457, in run
    use_user_site=options.use_user_site,
  File "/home/jsirois/.pex/pip.pex/3616f444e6841b3bb5e495d304f6de6a76f743d8/.deps/pip/pip/_internal/req/__init__.py", line 71, in install_given_reqs
    **kwargs
  File "/home/jsirois/.pex/pip.pex/3616f444e6841b3bb5e495d304f6de6a76f743d8/.deps/pip/pip/_internal/req/req_install.py", line 775, in install
    prefix=prefix,
  File "/home/jsirois/.pex/pip.pex/3616f444e6841b3bb5e495d304f6de6a76f743d8/.deps/pip/pip/_internal/locations.py", line 186, in get_scheme
    dist_name, user, home, root, isolated, prefix
  File "/home/jsirois/.pex/pip.pex/3616f444e6841b3bb5e495d304f6de6a76f743d8/.deps/pip/pip/_internal/locations.py", line 108, in distutils_scheme
    d = Distribution(dist_args)
  File "/home/jsirois/dev/pantsbuild/jsirois-pex/.tox/py27/lib/python2.7/site-packages/setuptools/dist.py", line 448, in __init__
    k: v for k, v in attrs.items()
  File "/usr/lib/python2.7/distutils/dist.py", line 287, in __init__
    self.finalize_options()
  File "/home/jsirois/dev/pantsbuild/jsirois-pex/.tox/py27/lib/python2.7/site-packages/setuptools/dist.py", line 740, in finalize_options
    ep.load()(self)
TypeError: unbound method _finalize_setup_keywords() must be called with Distribution instance as first argument (got Distribution instance instead)
Traceback (most recent call last):
  File "/usr/lib/python2.7/runpy.py", line 174, in _run_module_as_main
    "__main__", fname, loader, pkg_name)
  File "/usr/lib/python2.7/runpy.py", line 72, in _run_code
    exec code in run_globals
  File "/home/jsirois/dev/pantsbuild/jsirois-pex/pex/__main__.py", line 8, in <module>
    __name__ == '__main__' and pex.main()
  File "/home/jsirois/dev/pantsbuild/jsirois-pex/pex/bin/pex.py", line 724, in main
    pex_builder = build_pex(reqs, options, cache=ENV.PEX_ROOT)
  File "/home/jsirois/dev/pantsbuild/jsirois-pex/pex/bin/pex.py", line 644, in build_pex
    ignore_errors=options.ignore_errors)
  File "/home/jsirois/dev/pantsbuild/jsirois-pex/pex/resolver.py", line 824, in resolve_multi
    return list(resolve_request.resolve_distributions(ignore_errors=ignore_errors))
  File "/home/jsirois/dev/pantsbuild/jsirois-pex/pex/resolver.py", line 574, in resolve_distributions
    raise_type=Untranslateable):
  File "/home/jsirois/dev/pantsbuild/jsirois-pex/pex/resolver.py", line 392, in _run_parallel
    max_jobs=self._max_parallel_jobs
  File "/home/jsirois/dev/pantsbuild/jsirois-pex/pex/jobs.py", line 219, in execute_parallel
    raise error
pex.resolver.Untranslateable: pid: 45184 -> /home/jsirois/dev/pantsbuild/jsirois-pex/.tox/py27/bin/python /home/jsirois/.pex/pip.pex/3616f444e6841b3bb5e495d304f6de6a76f743d8 --disable-pip-version-check --isolated --no-python-version-warning -q --cache-dir /home/jsirois/.pex install --no-deps --no-index --only-binary :all: --target /home/jsirois/.pex/installed_wheels/4a13a89cdea76fef75d2f4f649fd2a71e63cadc5/my_app-0.0.0-py2-none-any.whl.b62ccffb18944c67b083ffc00365fcf3 --no-compile /home/jsirois/.pex/built_wheels/local_projects/tmpxEPsaM/2f24e7ea7df2645d261d6734d297043c0d527dfc/cp27-cp27mu/my_app-0.0.0-py2-none-any.whl raised Executing /home/jsirois/dev/pantsbuild/jsirois-pex/.tox/py27/bin/python /home/jsirois/.pex/pip.pex/3616f444e6841b3bb5e495d304f6de6a76f743d8 --disable-pip-version-check --isolated --no-python-version-warning -q --cache-dir /home/jsirois/.pex install --no-deps --no-index --only-binary :all: --target /home/jsirois/.pex/installed_wheels/4a13a89cdea76fef75d2f4f649fd2a71e63cadc5/my_app-0.0.0-py2-none-any.whl.b62ccffb18944c67b083ffc00365fcf3 --no-compile /home/jsirois/.pex/built_wheels/local_projects/tmpxEPsaM/2f24e7ea7df2645d261d6734d297043c0d527dfc/cp27-cp27mu/my_app-0.0.0-py2-none-any.whl failed with 2

FAILED

========================================================================================================================================== FAILURES ==========================================================================================================================================
___________________________________________________________________________________________________________________________________ test_entry_points_dict ___________________________________________________________________________________________________________________________________

    def test_entry_points_dict():
>     assert_entry_points({'console_scripts': ['my_app = my_app.my_module:do_something']})

tests/test_bdist_pex.py:54: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
tests/test_bdist_pex.py:37: in assert_entry_points
    with bdist_pex(project_dir) as my_app_pex:
/usr/lib/python2.7/contextlib.py:17: in __enter__
    return self.gen.next()
tests/test_bdist_pex.py:29: in bdist_pex
    spawn_python_job(args=cmd, cwd=project_dir, pythonpath=[pex_project_dir()]).wait()
.tox/py27/lib/python2.7/site-packages/pex/jobs.py:38: in wait
    self._check_returncode()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <pex.jobs.Job object at 0x7f932477f190>, stderr = None

    def _check_returncode(self, stderr=None):
      if self._process.returncode != 0:
        msg = 'Executing {} failed with {}'.format(' '.join(self._command), self._process.returncode)
        if stderr:
          msg += '\nSTDERR:\n{}'.format(stderr.decode('utf-8'))
>       raise self.Error(msg)
E       Error: Executing /home/jsirois/dev/pantsbuild/jsirois-pex/.tox/py27/bin/python -s setup.py --command-packages pex.commands bdist_pex --bdist-dir=/tmp/tmpsgDVu1 failed with 1

.tox/py27/lib/python2.7/site-packages/pex/jobs.py:67: Error
========================================================================================================================== 1 failed, 178 deselected in 3.08 seconds ==========================================================================================================================
ERROR: InvocationError for command /home/jsirois/dev/pantsbuild/jsirois-pex/.tox/py27/bin/pytest --ignore=tests/test_integration.py -vvsk test_entry_points_dict (exited with code 1)
__________________________________________________________________________________________________________________________________________ summary ___________________________________________________________________________________________________________________________________________
ERROR:   py27: commands failed
```

This is a tricky problem involving a custom importer virtualenv installs that loses references to symbols in its module after some import reloading by Pex, virtualenv, or both. Its unclear exactly where and when symbols get nuked at this point.

--- a/.travis.yml
+++ b/.travis.yml
@@ -1,13 +1,36 @@
+x-tox-env: &x-tox-env >
+  TOX_BIN=~/.tox-pex/bin
+  PATH="${TOX_BIN}:${PATH}"
+
+x-tox: &x-tox |
+  # Run tox from a self-contained virtual environment.
+  python3 -mvenv ~/.tox-pex
+  source ${TOX_BIN}/activate
+  pip install -U pip
+  pip install tox==3.14.1
+
+  # Prove we have the tox we just installed on the PATH.
+  which python
+  deactivate
+  tox --version
+
 x-linux-shard: &x-linux-shard
   os: linux
   dist: bionic
   language: python
-  install: |
-    pip install -U tox==3.14.1
+  install: *x-tox
+  env: *x-tox-env
+  addons:
+    apt:
+      packages:
+        - python3-venv
+
+# Pre-built Python distributions available on bionic (Ubuntu 18.04) are listed here:
+#   https://docs.travis-ci.com/user/languages/python/
 
 x-linux-27-shard: &x-linux-27-shard
   <<: *x-linux-shard
-  python: 2.7
+  python: 2.7.18
 
 x-linux-pypy-shard: &x-linux-pypy-shard
   <<: *x-linux-shard
@@ -15,48 +38,46 @@ x-linux-pypy-shard: &x-linux-pypy-shard
 
 x-linux-38-shard: &x-linux-38-shard
   <<: *x-linux-shard
-  python: 3.8
+  python: 3.8.2
+
+x-osx-env: &x-osx-env >
+  PYENV_ROOT="${HOME}/.pyenv_pex"
+  PATH="${PYENV_ROOT}/versions/${PYENV_VERSION}/bin:${PATH}"
 
 x-osx-shard: &x-osx-shard
   os: osx
   osx_image: xcode9.4
   language: generic
-  env: &env >
-    PYENV_ROOT="${HOME}/.pyenv_pex"
-    PATH="${PYENV_ROOT}/shims:${PATH}"
   cache:
     # The default is 3 minutes (180).
     timeout: 300
     directories:
       - .pyenv_test
       - "${PYENV_ROOT}"
-  install: |
-    PYENV="${PYENV_ROOT}/bin/pyenv"
-    if [ ! -x "${PYENV}" ]; then
-      rm -rf ${PYENV_ROOT}
-      git clone https://github.com/pyenv/pyenv "${PYENV_ROOT}";
-    fi
-    "${PYENV}" install --keep --skip-existing ${PYENV_VERSION}
-    "${PYENV}" global ${PYENV_VERSION}
-    pip install -U tox==3.14.1
-
-x-osx-ssl: &x-osx-ssl >
-  CPPFLAGS=-I/usr/local/opt/openssl/include
-  LDFLAGS=-L/usr/local/opt/openssl/lib
+  install:
+    - *x-tox
+    - |
+      # Ensure we have the targeted Python for tox to find on $PATH
+      PYENV="${PYENV_ROOT}/bin/pyenv"
+      if [ ! -x "${PYENV}" ]; then
+        rm -rf ${PYENV_ROOT}
+        git clone https://github.com/pyenv/pyenv "${PYENV_ROOT}";
+      fi
+      "${PYENV}" install --keep --skip-existing ${PYENV_VERSION}
 
 x-osx-27-shard: &x-osx-27-shard
   <<: *x-osx-shard
   env:
-    - *env
-    - *x-osx-ssl
-    - PYENV_VERSION=2.7.17
+    - *x-osx-env
+    - *x-tox-env
+    - PYENV_VERSION=2.7.18
 
 x-osx-38-shard: &x-osx-38-shard
   <<: *x-osx-shard
   env:
-    - *env
-    - *x-osx-ssl
-    - PYENV_VERSION=3.8.0
+    - *x-osx-env
+    - *x-tox-env
+    - PYENV_VERSION=3.8.2
 
 # NB: Travis partitions caches using a combination of os, language amd env vars. As such, we do not
 # use TOXENV and instead pass the toxenv via -e on the command line. This helps ensure we share
@@ -85,17 +106,17 @@ matrix:
 
     - <<: *x-linux-shard
       name: TOXENV=py35
-      python: 3.5
+      python: 3.5.9
       script: tox -v -e py35
 
     - <<: *x-linux-shard
       name: TOXENV=py36
-      python: 3.6
+      python: 3.6.10
       script: tox -v -e py36
 
     - <<: *x-linux-shard
       name: TOXENV=py37
-      python: 3.7
+      python: 3.7.7
       script: tox -v -e py37
 
     - <<: *x-linux-38-shard

--- a/tests/test_bdist_pex.py
+++ b/tests/test_bdist_pex.py
@@ -6,6 +6,7 @@ import subprocess
 from contextlib import contextmanager
 from textwrap import dedent
 
+from pex import resolver
 from pex.common import open_zip, temporary_dir
 from pex.interpreter import spawn_python_job
 from pex.testing import WheelBuilder, make_project, temporary_content
@@ -13,6 +14,31 @@ from pex.testing import WheelBuilder, make_project, temporary_content
 
 def pex_project_dir():
   return subprocess.check_output(['git', 'rev-parse', '--show-toplevel']).decode('utf-8').strip()
+
+
+BDIST_PEX_PYTHONPATH = None
+
+
+def bdist_pex_pythonpath():
+  # In order to run the bdist_pex distutils command we need:
+  # 1. setuptools on the PYTHONPATH since the test projects use and test setuptools.setup and its
+  #    additional features above and beyond distutils.core.setup like entry points declaration.
+  # 2. Pex on the PYTHONPATH so its distutils command module(s) can be found.
+  # 3. An indication to distutils of where to look for Pex distutils commands.
+  #
+  # We take care of 1 and 2 here and 3 is taken care of by passing --command-packages to distutils.
+
+  global BDIST_PEX_PYTHONPATH
+  if BDIST_PEX_PYTHONPATH is None:
+    BDIST_PEX_PYTHONPATH = [pex_project_dir()]
+
+    # Although the setuptools version is not important, we pick one so the test can leverage the
+    # pex cache for speed run over run.
+    BDIST_PEX_PYTHONPATH.extend(
+      resolved_distribution.distribution.location
+      for resolved_distribution in resolver.resolve(['setuptools==36.2.7'])
+    )
+  return BDIST_PEX_PYTHONPATH
 
 
 @contextmanager
@@ -26,7 +52,7 @@ def bdist_pex(project_dir, bdist_args=None):
     if bdist_args:
       cmd.extend(bdist_args)
 
-    spawn_python_job(args=cmd, cwd=project_dir, pythonpath=[pex_project_dir()]).wait()
+    spawn_python_job(args=cmd, cwd=project_dir, pythonpath=bdist_pex_pythonpath()).wait()
     dists = os.listdir(dist_dir)
     assert len(dists) == 1
     yield os.path.join(dist_dir, dists[0])

--- a/tox.ini
+++ b/tox.ini
@@ -6,6 +6,14 @@ skip_missing_interpreters = True
 # in `.travis.yml`.
 minversion = 3.14.1
 
+# N.B.: Pex tests run via tox for Python 2.7 fail starting at virtualenv==20.0.5 so we pin
+# virtualenv just below that and, correspondingly, tox at the highest version that supports
+# virtualenv 20.0.4.
+# See: https://github.com/pantsbuild/pex/issues/967
+requires =
+  tox<=3.14.5
+  virtualenv<20.0.5
+
 envlist =
 	py{py,27,38},style,isort-check,integration-tests
 

